src/realm/object-store/sync/app.hpp

```diff
@@ -245,9 +245,21 @@ class App : public std::enable_shared_from_this<App>,
         SharedApp m_parent;
     };
 
+    /// Retrieve a cached app instance if one was previously generated for `config`'s app_id+base_url combo,
+    /// otherwise generate and return a new instance and persist it in the cache.
     static SharedApp get_shared_app(const Config& config, const SyncClientConfig& sync_client_config);
+
+    /// Generate and return a new app instance for the given config, bypassing the app cache.
     static SharedApp get_uncached_app(const Config& config, const SyncClientConfig& sync_client_config);
-    static std::shared_ptr<App> get_cached_app(const std::string& app_id);
+
+    /// Return a cached app instance if one was previously generated for the `app_id`+`base_url` combo using
+    /// `get_shared_app`.
+    /// If base_url is not provided, and there are multiple cached apps with the same app_id but different base_urls,
+    /// then a random one will be returned.
+    ///
+    /// Prefer using `get_shared_app` to avoid the non-deterministic behavior.
+    static std::shared_ptr<App> get_cached_app(const std::string& app_id,
+                                               const util::Optional<std::string>* base_url = nullptr);
```

Comment:
this can  just be `const std::optional<std::string>& base_url = std::nullopt`

Reply:
I think we want to distinguish between an empty optional value vs the argument not being provided at all, which is not possible with the reference arg I think? My c++ is rusty so icbw 😅

Reply:
hm. i was trying to make a distinction between falling back to the default value or not to avoid any breaking changes, but i guess it's silly since it was not even possible to have a cache with multiple entries for the same app ID. i think i can simplify this without changing the existing behavior, will update 👍🏼